File: backend/vexel-rs/src/symmetry.rs
//! Stage 6b: symmetry of the placed boundary. Mirrors `symmetry.py`.

use crate::curves::P;

pub const SYM_MEAN: f64 = 0.10;
pub const SYM_MAX: f64 = 0.30;
pub const SYM_CAP: f64 = 1.0;
pub const MIN_VERTICES: usize = 24;
pub const ORDERS: [usize; 7] = [8, 7, 6, 5, 4, 3, 2];
pub const AXIS_SNAP_DEG: f64 = 1.5;

fn sub(a: P, b: P) -> P {
    [a[0] - b[0], a[1] - b[1]]
}

fn norm(a: P) -> f64 {
    (a[0] * a[0] + a[1] * a[1]).sqrt()
}

fn centroid(poly: &[P]) -> P {
    let n = poly.len() as f64;
    [poly.iter().map(|p| p[0]).sum::<f64>() / n, poly.iter().map(|p| p[1]).sum::<f64>() / n]
}

/// numpy's default (linear) percentile.
fn percentile(sorted: &[f64], q: f64) -> f64 {
    let n = sorted.len();
    if n == 0 {
        return 0.0;
    }
    let pos = (n as f64 - 1.0) * q / 100.0;
    let lo = pos.floor() as usize;
    let hi = pos.ceil() as usize;
    if lo == hi {
        sorted[lo]
    } else {
        sorted[lo] + (sorted[hi] - sorted[lo]) * (pos - lo as f64)
    }
}

/// Exact nearest-vertex lookup through a uniform grid: the search widens ring
/// by ring until no unsearched cell can hold a closer vertex.
pub(crate) struct Grid {
    cell: f64,
    x0: f64,
    y0: f64,
    nx: i64,
    ny: i64,
    bins: Vec<Vec<usize>>,
    pts: Vec<P>,
}

impl Grid {
    pub(crate) fn new(pts: &[P]) -> Grid {
        let cell = 2.0;
        let (mut x0, mut y0, mut x1, mut y1) = (f64::INFINITY, f64::INFINITY, f64::NEG_INFINITY, f64::NEG_INFINITY);
        for p in pts {
            x0 = x0.min(p[0]);
            y0 = y0.min(p[1]);
            x1 = x1.max(p[0]);
            y1 = y1.max(p[1]);
        }
        let nx = (((x1 - x0) / cell).floor() as i64 + 1).max(1);
        let ny = (((y1 - y0) / cell).floor() as i64 + 1).max(1);
        let mut bins = vec![Vec::new(); (nx * ny) as usize];
        for (i, p) in pts.iter().enumerate() {
            let cx = (((p[0] - x0) / cell).floor() as i64).clamp(0, nx - 1);
            let cy = (((p[1] - y0) / cell).floor() as i64).clamp(0, ny - 1);
            bins[(cy * nx + cx) as usize].push(i);
        }
        Grid { cell, x0, y0, nx, ny, bins, pts: pts.to_vec() }
    }

    pub(crate) fn nearest(&self, q: P) -> (f64, usize) {
        let cx = (((q[0] - self.x0) / self.cell).floor() as i64).clamp(0, self.nx - 1);
        let cy = (((q[1] - self.y0) / self.cell).floor() as i64).clamp(0, self.ny - 1);
        let mut best = (f64::INFINITY, 0usize);
        let max_ring = self.nx.max(self.ny);
        for ring in 0..=max_ring {
            // a vertex in a cell `ring` away is at least (ring - 1) cells off
            if ring > 0 && ((ring - 1) as f64) * self.cell > best.0 {
                break;
            }
            for dy in -ring..=ring {
                for dx in -ring..=ring {
                    if dx.abs() != ring && dy.abs() != ring {
                        continue;
                    }
                    let (gx, gy) = (cx + dx, cy + dy);
                    if gx < 0 || gy < 0 || gx >= self.nx || gy >= self.ny {
                        continue;
                    }
                    for &i in &self.bins[(gy * self.nx + gx) as usize] {
                        let d = norm(sub(self.pts[i], q));
                        if d < best.0 {
                            best = (d, i);
                        }
                    }
                }
            }
        }
        best
    }
}

/// (mean, worst, index of nearest vertex) for every image; see the Python `_match`.
fn matching(grid: &Grid, images: &[P]) -> (f64, f64, Vec<usize>) {
    let mut dist = Vec::with_capacity(images.len());
    let mut idx = Vec::with_capacity(images.len());
    for q in images {
        let (d, i) = grid.nearest(*q);
        dist.push(d);
        idx.push(i);
    }
    let mean = dist.iter().sum::<f64>() / dist.len() as f64;
    let max = dist.iter().cloned().fold(0.0f64, f64::max);
    let mut sorted = dist.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let p99 = percentile(&sorted, 99.0);
    let worst = if max <= SYM_CAP { p99 } else { max };
    (mean, worst, idx)
}

fn fits(mean: f64, worst: f64) -> bool {
    mean <= SYM_MEAN && worst <= SYM_MAX
}

pub fn reflect(p: P, c: P, d: P) -> P {
    // r = 2 d dᵀ − I
    let q = sub(p, c);
    let rx = (2.0 * d[0] * d[0] - 1.0) * q[0] + (2.0 * d[0] * d[1]) * q[1];
    let ry = (2.0 * d[0] * d[1]) * q[0] + (2.0 * d[1] * d[1] - 1.0) * q[1];
    [c[0] + rx, c[1] + ry]
}

pub fn rotate(p: P, c: P, angle: f64) -> P {
    let (ca, sa) = (angle.cos(), angle.sin());
    let q = sub(p, c);
    [c[0] + ca * q[0] - sa * q[1], c[1] + sa * q[0] + ca * q[1]]
}

/// Candidate mirror axes as (point on axis, unit direction). See the Python.
pub fn mirror_axes(poly: &[P]) -> Vec<(P, P)> {
    let c = centroid(poly);
    let n = poly.len().max(1) as f64;
    let (mut a, mut b, mut cc) = (0.0, 0.0, 0.0);
    for p in poly {
        let q = sub(*p, c);
        a += q[0] * q[0];
        b += q[0] * q[1];
        cc += q[1] * q[1];
    }
    a /= n;
    b /= n;
    cc /= n;
    let theta = 0.5 * (2.0 * b).atan2(a - cc);
    let d1 = [theta.cos(), theta.sin()];
    let d2 = [-theta.sin(), theta.cos()];
    let mut dirs: Vec<P> = vec![d1, d2, [d1[0] + d2[0], d1[1] + d2[1]], [d1[0] - d2[0], d1[1] - d2[1]]];
    for k in 0..12 {
        let ang = (15.0 * k as f64).to_radians();
        dirs.push([ang.cos(), ang.sin()]);
    }
    let mut out: Vec<(P, P)> = Vec::new();
    let cos2 = 2.0f64.to_radians().cos();
    for d in dirs {
        let len = norm(d);
        if len < 1e-12 {
            continue;
        }
        let mut d = [d[0] / len, d[1] / len];
        if d[0] < 0.0 || (d[0] == 0.0 && d[1] < 0.0) {
            d = [-d[0], -d[1]];
        }
        let ang = d[1].atan2(d[0]).to_degrees();
        if ang.abs().min((180.0 - ang).abs()) <= AXIS_SNAP_DEG {
            d = [1.0, 0.0];
        } else if (ang - 90.0).abs() <= AXIS_SNAP_DEG {
            d = [0.0, 1.0];
        }
        if out.iter().all(|(_c, e)| (d[0] * e[0] + d[1] * e[1]).abs() < cos2) {
            out.push((c, d));
        }
    }
    out
}

/// The ring made exactly mirror-symmetric about `axis`, or None when it is not.
pub fn symmetrize(poly: &[P], axis: (P, P)) -> Option<Vec<P>> {
    let (c, d) = axis;
    let grid = Grid::new(poly);
    let images: Vec<P> = poly.iter().map(|p| reflect(*p, c, d)).collect();
    let (mean, worst, idx) = matching(&grid, &images);
    if !fits(mean, worst) {
        return None;
    }
    Some(
        poly.iter()
            .zip(idx)
            .map(|(p, i)| {
                let m = reflect(poly[i], c, d);
                [0.5 * (p[0] + m[0]), 0.5 * (p[1] + m[1])]
            })
            .collect(),
    )
}

pub fn rotational_order(poly: &[P]) -> usize {
    let c = centroid(poly);
    let grid = Grid::new(poly);
    for &k in ORDERS.iter() {
        let angle = 2.0 * std::f64::consts::PI / k as f64;
        let images: Vec<P> = poly.iter().map(|p| rotate(*p, c, angle)).collect();
        let (mean, worst, _idx) = matching(&grid, &images);
        if fits(mean, worst) {
            return k;
        }
    }
    1
}

pub fn symmetrize_rotational(poly: &[P], k: usize) -> Vec<P> {
    let c = centroid(poly);
    let grid = Grid::new(poly);
    let mut acc: Vec<P> = poly.to_vec();
    for j in 1..k {
        let angle = 2.0 * std::f64::consts::PI * j as f64 / k as f64;
        let images: Vec<P> = poly.iter().map(|p| rotate(*p, c, angle)).collect();
        let (_m, _w, idx) = matching(&grid, &images);
        for (a, &i) in acc.iter_mut().zip(idx.iter()) {
            let back = rotate(poly[i], c, -angle);
            a[0] += back[0];
            a[1] += back[1];
        }
    }
    let kf = k as f64;
    acc.iter().map(|a| [a[0] / kf, a[1] / kf]).collect()
}

/// (the symmetrised ring or None, the mirror axes that fit, best first). See the Python.
pub fn ring_symmetries(poly: &[P]) -> (Option<Vec<P>>, Vec<(P, P)>) {
    if poly.len() < MIN_VERTICES {
        return (None, Vec::new());
    }
    let mut out: Vec<P> = poly.to_vec();
    let mut changed = false;
    let k = rotational_order(&out);
    if k > 1 {
        out = symmetrize_rotational(&out, k);
        changed = true;
    }
    let mut axes: Vec<(f64, (P, P))> = Vec::new();
    for axis in mirror_axes(&out) {
        if let Some(sym) = symmetrize(&out, axis) {
            let grid = Grid::new(&out);
            let images: Vec<P> = out.iter().map(|p| reflect(*p, axis.0, axis.1)).collect();
            let (mean, _w, _i) = matching(&grid, &images);
            axes.push((mean, axis));
            out = sym;
            changed = true;
        }
    }
    // best first by a rule both implementations agree on to the last bit
    let rank = |d: P| -> (u8, f64) {
        let exact = if (d[0] == 1.0 && d[1] == 0.0) || (d[0] == 0.0 && d[1] == 1.0) { 0 } else { 1 };
        (exact, d[1].atan2(d[0]).to_degrees().rem_euclid(180.0))
    };
    axes.sort_by(|a, b| {
        let (ra, rb) = (rank(a.1 .1), rank(b.1 .1));
        ra.0.cmp(&rb.0).then(ra.1.partial_cmp(&rb.1).unwrap())
    });
    (if changed { Some(out) } else { None }, axes.into_iter().map(|(_m, a)| a).collect())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn regular(n_sides: usize, r: f64, step: f64) -> Vec<P> {
        let mut pts = Vec::new();
        let mut seed: u64 = 7;
        let mut noise = || {
            seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
            ((seed >> 33) as f64 / (1u64 << 31) as f64 - 0.5) * 0.08
        };
        for k in 0..n_sides {
            let a0 = 2.0 * std::f64::consts::PI * k as f64 / n_sides as f64;
            let a1 = 2.0 * std::f64::consts::PI * (k + 1) as f64 / n_sides as f64;
            let p0 = [128.0 + r * a0.cos(), 128.0 + r * a0.sin()];
            let p1 = [128.0 + r * a1.cos(), 128.0 + r * a1.sin()];
            let m = (norm(sub(p1, p0)) / step) as usize;
            for i in 0..m {
                let f = i as f64 / m as f64;
                pts.push([p0[0] + f * (p1[0] - p0[0]) + noise(), p0[1] + f * (p1[1] - p0[1]) + noise()]);
            }
        }
        pts
    }

    #[test]
    fn a_hexagon_has_order_six_and_a_square_four_mirror_axes() {
        let hexagon = regular(6, 80.0, 0.7);
        assert_eq!(rotational_order(&hexagon), 6);
        let out = symmetrize_rotational(&hexagon, 6);
        let c = centroid(&out);
        let grid = Grid::new(&out);
        let turned: Vec<P> = out.iter().map(|p| rotate(*p, c, std::f64::consts::PI / 3.0)).collect();
        let (_m, worst, _i) = matching(&grid, &turned);
        assert!(worst < 0.05, "{worst}");
        let square = regular(4, 80.0, 0.7);
        let hits = mirror_axes(&square).into_iter().filter(|ax| symmetrize(&square, *ax).is_some()).count();
        assert!(hits >= 4, "{hits}");
        assert_eq!(rotational_order(&regular(5, 80.0, 0.7)), 5);
    }

    #[test]
    fn the_grid_finds_the_true_nearest_vertex() {
        let pts = regular(3, 60.0, 0.9);
        let grid = Grid::new(&pts);
        for q in [[100.0, 100.0], [128.0, 68.0], [190.0, 160.0], [0.0, 0.0]] {
            let (d, i) = grid.nearest(q);
            let brute = pts.iter().map(|p| norm(sub(*p, q))).fold(f64::INFINITY, f64::min);
            assert!((d - brute).abs() < 1e-12 && (norm(sub(pts[i], q)) - brute).abs() < 1e-12);
        }
    }
}

File: backend/studi0trace/engines/vexel/symmetry.py
"""Stage 6b: symmetry of the placed boundary.

A mark that is mirror-symmetric, or the same under a turn of 360°/k, was drawn
so, and a trace that has one side a tenth of a pixel fatter than the other
reads as hand-made even when every edge is inside tolerance. Before the nodes
are placed and the curves fitted, each region's ring is tested against its
candidate symmetries: reflect (or turn) every vertex and ask how far it lands
from the nearest vertex of the ring. When the mean is inside SYM_MEAN and the
worst inside SYM_MAX — the placement's own uncertainty — the symmetry is
real, and every vertex is replaced by the average of itself and its images,
so the ring becomes exactly symmetric. The vertices live in the shared arcs,
so the neighbour on the other side of each edge moves with it.

Vectorizer.AI lists this under "symmetry modelling". The candidate axes are
the ring's principal directions, their 45° rotations, and every 15° about the
centroid, which covers the regular polygons whose covariance is isotropic.
"""
from __future__ import annotations

import math

import numpy as np
from scipy.spatial import cKDTree

SYM_MEAN = 0.10      # px; mean distance from each image to the nearest vertex
SYM_MAX = 0.30       # px; the 99th-percentile image — one cusp vertex handed back a sliver does not decide
SYM_CAP = 1.0        # px; but nothing may be further off than this
MIN_VERTICES = 24    # a ring shorter than this is too little to judge
ORDERS = (8, 7, 6, 5, 4, 3, 2)
AXIS_SNAP_DEG = 1.5  # a mirror axis this close to vertical or horizontal is exactly so
# Two vertices this close to equally near an image are equally near: the one
# with the lower index is its match. On a ring placed on the pixel lattice an
# image often lands exactly between two vertices, and which one a k-d tree (or
# the Rust grid) found first decided where the averaged vertex went.
NEAR_TIE = 1e-9
# A ring whose covariance is this close to isotropic (relative to its trace)
# has no principal direction: every direction is an eigenvector, and the one
# an eigensolver hands back is decided by the last bits of the sums (a square
# made exactly four-fold symmetric is one). Its candidates are the 15° grid's.
ISOTROPIC = 1e-9


def _centroid(poly: np.ndarray) -> np.ndarray:
    return poly.mean(axis=0)


def _nearest(poly: np.ndarray, images: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """(distance, index) of the nearest vertex to every image, a tie (within
    NEAR_TIE) going to the lower index."""
    tree = cKDTree(poly)
    k = min(3, len(poly))
    dist, idx = tree.query(images, k=k)
    if k == 1:
        return dist, idx
    tied = dist <= dist[:, :1] + NEAR_TIE
    best = np.where(tied, idx, len(poly)).min(axis=1)
    # all k returned are tied: there may be more beyond them
    for j in np.nonzero(tied[:, -1])[0]:
        best[j] = min(tree.query_ball_point(images[j], float(dist[j, 0]) + NEAR_TIE))
    return dist[:, 0], best


def _match(poly: np.ndarray, images: np.ndarray) -> tuple[float, float, np.ndarray]:
    """(mean, worst, index of nearest vertex) for every image; `worst` is the
    99th percentile with the absolute maximum folded in through SYM_CAP."""
    dist, idx = _nearest(poly, images)
    p99 = float(np.percentile(dist, 99))
    worst = p99 if float(dist.max()) <= SYM_CAP else float(dist.max())
    return float(dist.mean()), worst, idx


def _fits(mean: float, worst: float) -> bool:
    return mean <= SYM_MEAN and worst <= SYM_MAX


def mirror_axes(poly: np.ndarray) -> list[tuple[np.ndarray, np.ndarray]]:
    """Candidate mirror axes as (point on axis, unit direction)."""
    c = _centroid(poly)
    q = poly - c
    n = max(len(q), 1)
    a, b, cc = float(q[:, 0] @ q[:, 0]) / n, float(q[:, 0] @ q[:, 1]) / n, float(q[:, 1] @ q[:, 1]) / n
    dirs: list[np.ndarray] = []
    # the principal directions in closed form, the major one first, with the
    # minor one a quarter turn anticlockwise of it (an eigensolver's signs are
    # its own, and they decided which diagonal came first)
    if math.hypot(a - cc, 2.0 * b) > ISOTROPIC * (a + cc):
        theta = 0.5 * math.atan2(2.0 * b, a - cc)
        d1 = np.array([math.cos(theta), math.sin(theta)])
        d2 = np.array([-math.sin(theta), math.cos(theta)])
        dirs = [d1, d2, d1 + d2, d1 - d2]
    dirs += [np.array([math.cos(math.radians(a)), math.sin(math.radians(a))]) for a in range(0, 180, 15)]
    out: list[tuple[np.ndarray, np.ndarray]] = []
    for d in dirs:
        n = float(np.linalg.norm(d))
        if n < 1e-12:
            continue
        d = d / n
        if d[0] < 0 or (d[0] == 0 and d[1] < 0):
            d = -d
        # an axis within AXIS_SNAP_DEG of the canvas axes is one: a symmetric
        # mark is drawn upright, and the exact direction reflects exactly
        ang = math.degrees(math.atan2(d[1], d[0]))
        if min(abs(ang), abs(180.0 - ang)) <= AXIS_SNAP_DEG:
            d = np.array([1.0, 0.0])
        elif abs(ang - 90.0) <= AXIS_SNAP_DEG:
            d = np.array([0.0, 1.0])
        if all(abs(float(d @ e)) < math.cos(math.radians(2.0)) for _c, e in out):
            out.append((c, d))
    return out


def reflect(pts: np.ndarray, c: np.ndarray, d: np.ndarray) -> np.ndarray:
    r = 2.0 * np.outer(d, d) - np.eye(2)
    return c + (pts - c) @ r


def rotate(pts: np.ndarray, c: np.ndarray, angle: float) -> np.ndarray:
    ca, sa = math.cos(angle), math.sin(angle)
    r = np.array([[ca, -sa], [sa, ca]])
    return c + (pts - c) @ r.T


def symmetrize(poly: np.ndarray, axis: tuple[np.ndarray, np.ndarray]) -> np.ndarray | None:
    """The ring made exactly mirror-symmetric about `axis`, or None when it is not."""
    c, d = axis
    images = reflect(poly, c, d)
    mean, worst, idx = _match(poly, images)
    if not _fits(mean, worst):
        return None
    return 0.5 * (poly + reflect(poly[idx], c, d))


def rotational_order(poly: np.ndarray) -> int:
    """The largest k in ORDERS under which the ring maps onto itself, else 1."""
    c = _centroid(poly)
    for k in ORDERS:
        mean, worst, _idx = _match(poly, rotate(poly, c, 2.0 * math.pi / k))
        if _fits(mean, worst):
            return k
    return 1


def symmetrize_rotational(poly: np.ndarray, k: int) -> np.ndarray:
    """Every vertex averaged with its k−1 images turned back into place."""
    c = _centroid(poly)
    acc = poly.copy()
    for j in range(1, k):
        angle = 2.0 * math.pi * j / k
        _mean, _worst, idx = _match(poly, rotate(poly, c, angle))
        acc += rotate(poly[idx], c, -angle)
    return acc / k


def ring_symmetries(poly: np.ndarray) -> tuple[np.ndarray | None, list[tuple[np.ndarray, np.ndarray]]]:
    """Apply every symmetry the ring has: the rotational order first, then each
    mirror axis that fits. Returns (the symmetrised ring or None, the mirror
    axes that fit, best first)."""
    if len(poly) < MIN_VERTICES:
        return None, []
    out = poly
    changed = False
    k = rotational_order(out)
    if k > 1:
        out = symmetrize_rotational(out, k)
        changed = True
    axes: list[tuple[float, tuple[np.ndarray, np.ndarray]]] = []
    for axis in mirror_axes(out):
        sym = symmetrize(out, axis)
        if sym is not None:
            mean, _worst, _idx = _match(out, reflect(out, *axis))
            axes.append((mean, axis))
            out = sym
            changed = True
    # best first by a rule both implementations agree on to the last bit: an
    # exact canvas-axis direction first, then the smaller angle
    axes.sort(key=lambda t: _axis_rank(t[1]))
    return (out if changed else None), [a for _m, a in axes]


def _axis_rank(axis: tuple[np.ndarray, np.ndarray]) -> tuple[int, float]:
    d = axis[1]
    exact = 0 if (d[0] == 1.0 and d[1] == 0.0) or (d[0] == 0.0 and d[1] == 1.0) else 1
    return exact, math.degrees(math.atan2(d[1], d[0])) % 180.0


def symmetrize_ring(poly: np.ndarray) -> np.ndarray | None:
    """The ring with every symmetry it has applied, or None when it has none."""
    return ring_symmetries(poly)[0]
